### DISCUSSION TITLE
.net 7 library project errors

### DISCUSSION THREAD

**Discussion by sam-wheat** [unoplatform/uno#10811] [Q&A] [answered]
I have a library project and host project.  Each builds by itself.  When I reference the library from the host I get errors:

> Error	CS0433	The type 'Application' exists in both 'Microsoft.WinUI, Version=3.0.0.0, Culture=neutral, PublicKeyToken=de31ebe4ad15742b' and 'Uno.UI, Version=255.255.255.255, Culture=neutral, PublicKeyToken=null'	

> Error	CS0433	The type 'Frame' exists in both 'Microsoft.WinUI, Version=3.0.0.0, Culture=neutral, PublicKeyToken=de31ebe4ad15742b' and 'Uno.UI, Version=255.255.255.255, Culture=neutral, PublicKeyToken=null'	

I added `net7.0;` to the `TargetFrameworks` section of the library project.  

    <TargetFrameworks>net7.0;net7.0-windows;</TargetFrameworks>

When I remove it I get (and other errors):

> Error		Project '..\' targets 'net7.0-windows;net7.0-ios;net7.0-macos;net7.0-maccatalyst;net7.0-android'. It cannot be referenced by a project that targets '.NETCoreApp,Version=v7.0'.		C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Current\Bin\amd64\Microsoft.Common.CurrentVersion.targets	1830	

How can I fix this?

**Comment by jeromelaban** [comment]
Thank you for the report. Are you able to create and attach here a sample project with what you're trying to do?

**Reply by sam-wheat** [reply]
Hi @jeromelaban 

Please see [this repo](https://github.com/leaderanalytics/LeaderPivot.XAML.UNO).

**Reply by jeromelaban** [reply]
Thanks for the update. To build for WinAppSDK, you'll need to specify the min SDK version in the target framework `net7.0-windows10.0.18362`, in this case, and not `net7.0-windows`.

If you change the `TargetFramework` to a different value, you'll also need to change the `TargetFramework` values in the rest of the file.

**Reply by sam-wheat** [reply]
Thank you @jeromelaban - I continue to have problems so I deleted and recreated everything.

Host project builds (prior to adding reference to lib).
Library project builds.
After adding the reference I get:


> Error	WMC0001	Unknown type 'LeaderPivotControl' in XML namespace 'clr-namespace:LeaderAnalytics.LeaderPivot.XAML.Uno;assembly=LeaderAnalytics.LeaderPivot.XAML.Uno'	LeaderPivot.XAML.Uno.Host.Windows	C:\Git\LeaderPivot.XAML.UNO\LeaderPivot.XAML.Uno.Host\LeaderPivot.XAML.Uno.Host.Shared\MainPage.xaml


> Error	XLS0414	The type 'ThemeResource' was not found. Verify that you are not missing an assembly reference and that all referenced assemblies have been built.	LeaderPivot.XAML.Uno.Host.Wasm	C:\Git\LeaderPivot.XAML.UNO\LeaderPivot.XAML.Uno.Host\LeaderPivot.XAML.Uno.Host.Shared\MainPage.xaml

I tried:
Changing TargetFramework from 18362 to 19041 in the library project
Changing TargetFramework from 19041 to 18362 in the host project
Deleting bin/obj folders, restarting VS (did this many times since I get often get different error messages).


The repo as it stands right now is the out-of-the-box templates for all projects.  In the library project I replaced net6 with net7 and added nuget metadata.

**Reply by sam-wheat** [reply]
@ttustonic 

Regarding [this thread](https://github.com/unoplatform/uno/discussions/10649) - I tried changing the versions per your link however it does not work.   I also tried changing the .csproj as follows with no luck:

        <ItemGroup>
		<PackageReference Include="Microsoft.WindowsAppSDK" Version="1.2.221109.1" />
		<PackageReference Include="Microsoft.Windows.SDK.BuildTools" Version="10.0.22621.755" />
		<PackageReference Include="Microsoft.Extensions.Logging.Console" Version="5.0.0" />
		<PackageReference Include="Uno.Core.Extensions.Logging.Singleton" Version="4.0.1" />
		<PackageReference Include="Uno.WinUI" Version="4.7.0-dev.417" />


		<Manifest Include="$(ApplicationManifest)" />
        </ItemGroup>


        <ItemGroup>
		
		<FrameworkReference Update="Microsoft.Windows.SDK.NET.Ref" RuntimeFrameworkVersion="10.0.22000.27" />
		<FrameworkReference Update="Microsoft.Windows.SDK.NET.Ref" TargetingPackVersion="10.0.22000.27" />
        </ItemGroup>

**Comment by sam-wheat** [comment, marked as answer]
Guys I am still blocked on this.   I have tried everything that seems remotely logical and I'm still stuck.  An assist is greatly appreciated.  Thank you!

**Reply by jeromelaban** [reply]
When a project does not build on windows, it's not something that Uno can fix explicitly.

In this case, your issue is here:

https://github.com/leaderanalytics/LeaderPivot.XAML.UNO/blob/6818e3e1a5d54e5c16f499765099152525ef30ef/LeaderPivot.XAML.Uno.Host/LeaderPivot.XAML.Uno.Host.Shared/MainPage.xaml#L8

Where using the `clr-namespace:` is not supported by WinAppSDK (This is a WPF only syntax). Use the [`using:` syntax instead](https://learn.microsoft.com/en-us/windows/uwp/xaml-platform/xaml-namespaces-and-namespace-mapping#mapping-custom-types-to-xaml-namespaces-and-prefixes).

**Reply by jeromelaban** [reply]
Thanks for the test. What does not build? Can you [provide a binlog](https://platform.uno/docs/articles/uno-builds-troubleshooting.html#generating-msbuild-binary-log-files)?

**Reply by sam-wheat** [reply]
The control library builds - problem is windows / Wasm / etc projects can't reference it.

> Error	WMC0001	Unknown type 'LeaderPivotControl' in XML namespace 'using:LeaderAnalytics.LeaderPivot.XAML.Uno;'

[LeaderPivot.XAML.Uno.Host.Windows_Debug_x64_Build_2023-01-10T19_35_53.9622949-08_00.zip](https://github.com/unoplatform/uno/files/10388765/LeaderPivot.XAML.Uno.Host.Windows_Debug_x64_Build_2023-01-10T19_35_53.9622949-08_00.zip)

**Reply by jeromelaban** [reply]
Thanks. The [using: syntax](https://learn.microsoft.com/en-us/windows/uwp/xaml-platform/xaml-namespaces-and-namespace-mapping#mapping-custom-types-to-xaml-namespaces-and-prefixes) does not allow for a `;` at the end, if you remove it, it will build further.

**Reply by sam-wheat** [reply]
Good catch - removed the semicolon. 
The project builds and runs but the control does not render.  No errors, output messages, binding errors that I can see.  Only VS intellisense complains.

**Reply by jeromelaban** [reply]
Thanks. Templatable controls are generally inheriting from `Control`, not `UserControl`. See [this sample](https://github.com/unoplatform/Uno.Samples/tree/master/UI/ControlLibrary) for details.

**Reply by sam-wheat** [reply]
I did 50 pushups.  Not all at once... but did 50.

**Reply by sam-wheat** [reply]
@jeromelaban Thanks for going above and beyond the call of duty. I was convinced I had version problem somewhere - was not even looking at the obvious.  Problem solved.

**Reply by jeromelaban** [reply]
Thanks for the update!

**Comment by leaderanalytics** [comment]
@jeromelaban Thank you - I made the suggested change and checked it in.  It still does not build.
